_posts/2015-10-22-transcending-rest-and-rpc.md

```diff
@@ -0,0 +1,102 @@
+---
+layout: post
+name: Transcending-rest-and-rpc
+title: 'Transcending REST and RPC'
+date: 2015-10-22 00:20:00 +00:00
+author: Felipe Fernandez
+image:
+    src: /assets/img/custom/blog/falcor.jpg
+tags:
+- rest
+- rpc
+- microservices
+
+---
+It seems that a new paradigm is coming. Facebook and Netflix have come with different implementations for that idea. Some people are calling it Demand Driven Architecture, but before I propose some solutions, let's go over some history to understand the problem. I will use the example that Netflix provides, but I think that most of us will find the patterns familiar.
+
+Just for clarity, let's assume that we have an on-demand streaming service -- quite similar to Netflix :) -- that has three microservices. One contains the genres, another one the titles and the last one contains the ratings that different users give to those titles. As we're in 2015 we'll discard RPC as architectural style and we'll go with a RESTful design. Let's briefly enumerate the characteristics of a RESTful design:
+
+* HTTP verbs have direct correspondence with CRUD verbs. One of the benefits of that is that clients can understand what is going on the server without digging into the implementation. For instance, we know that PUT is idempotent, so as there will not be any side effect, the client can call to that endpoint several times in a row without any issue.
+* We use extensive HTTP caching. Through headers like Cache-Control or ETag, we leverage one of the most juicy features of HTTP: caching resources.
+* HATEOAS for the win. Every single representation should have links to different representations so the clients can navigate through our backend like a graph.
+
+## REST limitations
+
+Let's imagine that we do a GET over genres endpoint. The response will include links to other GET requests like titles/1234 or ratings/6789. If we have Titanic title in different genres (romance, drama and DiCaprioRules) you can see how the client is going to do at least two redundant requests. Apart of that the client is going to reshape the data whenever comes back since it's likely that the client won't need every single field that those responses include. So, we have two problems with this design:
```

Comment:
send a GET to the genres endpoint

If we have the title "Titanic" in different genres

Apart from that

the client is going to reshape the date whenever it comes back